audacity/audacity
# Time stretching 2 of 6 make audio track stretching effective

Resolves: #4850

This PR enables the time-stretching of audio clips!
Changing the project tempo _will_ play the clips slower or faster. Clips can also individually be stretched using the stretch handle (like trimming, but with alt depressed).

Depends on
- #5040
- #5095




Recommended:

## QA:
This is a big change, with many consequences, please plan some time to play around and do all sorts of crazy things with it. I'd be surprised if you don't find bugs, other than those underneath that will get fixed in follow-up PRs:
* drawing tools won't work on stretched clips
* copy/paste/join actions involving stretched clips
* applying effects on stretched clips.
* generating sounds on stretched clips

Checklist:
- [x] No regression of the fix for this bug https://bugzilla.audacityteam.org/show_bug.cgi?id=2288
- [x] Playback works in all directions (also backward by scrubbing)
- [x] Various spectrum displays (standard, autocorrelation, reassignment, ...)
- [x] Reassignment algorithm of spectrograms especially -- stretch clip while in that view, see that the display updates properly.  Use an example where the difference between Reassignment and the default algorithm is noticeable.
- [x] Playback of stretched clips
- [x] Scrubbing including backward and forwards, and at ends of stretched clips
- [x] Correct playback with an envelope in a stretched clip (edit the envelope before you stretch)
- [x] Contrast tool:
  - Make two clips on a track with different stretch
  - Select all, note RMS values
  - Mix and render to new track
  - Repeat, observe the same RMS
 - [x] Playback stops correctly when it is started from a position other than 0.
 - [x] Persistency of stretches in the saved project file
 - [x] Rendering and exporting of stretches
 - [x] Rebase onto #5095 and confirm that the problem does not recur
 - [x] Change project tempo with a stereo track present, then split it to mono, observe correct stretch in new track
 - [x] Undo and Redo of changes to project tempo with all types of tracks present

## Issues  that need to be addressed within this PR:

- [x] Deadlock when using "At Zero Crossings"
- [x] Stretched clips are quieter
- [x] The right channel of a stereo track seems to ignore the envelope points
- [x] Beat Finder fails when analyzing a stretched (shrunk) clip
- [x] Beat Finder may cause deadlock
- [x] Insignificant lag when playing a stretched clip not from the zero mark
- [x] Deadlock when playing extremely stretched-out clips
- [x] Poor processing of synthetic tones resulting in high peaks (TBD)
- [x] Lag when playing a clip shrunk to extreme minimum
